File: AlgoPersonalStudy/src/backjoon/gold/BJ_17141.java
package backjoon.gold;

import java.io.*;
import java.util.*;
public class BJ_17141 {
    static int N = 0, voidCount = 0, answer = Integer.MAX_VALUE;
    static int[] deltaX = {1, 0, -1, 0}, deltaY = {0,1,0,-1};
    static int[][] map;
    static boolean success = false;
    static boolean[] checkedVirus;
    static List<int[]> virus = new ArrayList<>();;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        map = new int[N][N];

        int virusCount = 0;
        for(int n=0;n<N;n++){
            st = new StringTokenizer(br.readLine());
            for(int n2=0;n2<N;n2++){
                int num = Integer.parseInt(st.nextToken());
                map[n][n2] = num;
                if(num == 2) {
                    virus.add(new int[]{n,n2});
                    virusCount++;
                }
                else if(num == 1) voidCount++;
            }
        }

        if(voidCount == N*N) System.out.println(0);
        else {
            checkedVirus = new boolean[virusCount];
            getPosition(virusCount, 0, 0, M);
            System.out.println(success ? answer : -1);
        }
    }
    static void getPosition(int virusCount, int start, int dep, int position){
        if(position == dep){
            int checkedPosition = 0;
            int[] virusPosition = new int[position];
            for(int i=0;i<checkedVirus.length;i++){
                if(checkedVirus[i]){
                    virusPosition[checkedPosition++] = i;
                }
            }
            if(BFS(virusPosition)) success = true;
        }
        else{
            for(int i=start;i<virusCount;i++){
                if(!checkedVirus[i]){
                    checkedVirus[i] = true;
                    getPosition(virusCount, i+1, dep+1, position);
                    checkedVirus[i] = false;
                }
            }
        }
    }
    static boolean BFS(int[] virusPosition){
        Queue<int[]> labs = new LinkedList<>();
        int count = -1;
        boolean success = false;
        boolean[][] visited = new boolean[N][N];

        for(int index : virusPosition){
            int[] tmpPosition = virus.get(index);
            labs.add(tmpPosition);
        }

        while(!labs.isEmpty()){
            int qLen = labs.size();
            boolean working = false;
            for(int l=0;l<qLen;l++){
                int[] point = labs.poll();
                if(!visited[point[0]][point[1]]){
                    visited[point[0]][point[1]] = true;
                    working = true;
                    for (int d = 0; d < 4; d++) {
                        int dx = deltaX[d] + point[1];
                        int dy = deltaY[d] + point[0];
                        if (dx < 0 || dx >= N || dy < 0 || dy >= N) continue;
                        if (visited[dy][dx] || map[dy][dx] == 1) continue;
                        labs.add(new int[]{dy, dx});
                    }
                }
            }
            if(working) count++;
        }
        if(checkSuccess(visited)) {
            success = true;
            answer = Math.min(answer, count);
        }
        return success;
    }
    static boolean checkSuccess(boolean[][] lab){
        int tmpCount = 0;
        for(boolean[] col : lab){
            for(boolean row : col){
                if(!row) tmpCount++;
            }
        }
        return tmpCount == voidCount;
    }
}

File: AlgoPersonalStudy/src/_kotlin/backjoon/gold/BJ_17141.kt
package _kotlin.backjoon.gold

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

class BJ_17141 {
    var n = 0
    private var voidCount = 0
    var answer = Integer.MAX_VALUE
    val deltaX = arrayOf(1, 0, -1, 0)
    val deltaY = arrayOf(0, 1, 0, -1)
    var map:Array<Array<Int>>? = null
    private var success:Boolean = false
    private var checkedVirus: BooleanArray? = null
    private val virus = ArrayList<IntArray>()

    fun solution(){
        val br = BufferedReader(InputStreamReader(System.`in`))
        val (tmpN, m) = br.readLine().split(" ").map { it.toInt() }
        n = tmpN

        map = Array(n){Array(n){0} }
        var virusCount = 0

        for(y:Int in 0 until n){
            val st = StringTokenizer(br.readLine())
            for(x:Int in 0 until n){
                val num = Integer.parseInt(st.nextToken())
                map!![y][x] = num
                if(num == 2){
                    virus.add(intArrayOf(y,x))
                    virusCount++
                }
                else if(num == 1) voidCount++
            }
        }
        if(voidCount == n*n) println(0)
        else{
            checkedVirus = BooleanArray(virusCount)
            getPosition(virusCount, 0, 0, m)
            println(if(success) 0 else -1)

        }
    }
    private fun getPosition(virusCount:Int, start:Int, dep:Int, position:Int) {
        if(position == dep){
            var checkedPosition = 0
            val virusPosition = IntArray(position)
            for(i:Int in 0 until (checkedVirus?.size ?: 0)){
                if(checkedVirus!![i]){
                    virusPosition[checkedPosition++] = i
                }
            }
            if(BFS(virusPosition)) success = true
        }
        else{
            for(i:Int in start until virusCount){
                if(!checkedVirus!![i]){
                    checkedVirus!![i] = true
                    getPosition(virusCount, i+1, dep+1, position)
                    checkedVirus!![i] = false
                }
            }
        }
    }
    private fun BFS(virusPosition: IntArray):Boolean {
        val labs = LinkedList<IntArray>()
        var count = -1
        var success = false

        val visited = Array(n){BooleanArray(n)}
        for(index:Int in virusPosition){
            val tmpPosition = virus[index]
            labs.add(tmpPosition)
        }

        while(!labs.isEmpty()){
            val labSize = labs.size
            var working = false
            for(l:Int in 0 until labSize){
                val point = labs.poll()
                if(!visited[point[0]][point[1]]){
                    visited[point[0]][point[1]] = true
                    working = true
                    for(d:Int in 0 until 4){
                        val dx = deltaX[d] + point[1]
                        val dy = deltaY[d] + point[0]
                        if(dx < 0 || dx >= n || dy < 0 || dy >= n) continue
                        if(visited[dy][dx] || map!![dy][dx] == 1) continue
                        labs.add(intArrayOf(dy, dx))
                    }
                }
            }
            if(working) count++
        }
        if(checkSuccess(visited)){
            success = true
            answer = Math.min(answer, count)
        }
        return success
    }
    private fun checkSuccess(visited:Array<BooleanArray>):Boolean{
        var tmpCount = 0
        for(col in visited){
            for(row in col){
                if(!row) tmpCount++
            }
        }
        return tmpCount == voidCount
    }
}
fun main(){
    val solutionClass = BJ_17141()
    solutionClass.solution()
}